File: Algorithms/src/1823. Find the Winner of the Circular Game/Code/solution.java
class Solution {
    public int findTheWinner(int n, int k) {
        // Calls the recursive Josephus function and adjusts the result from 0-indexed
        // to 1-indexed
        return josephus(n, k) + 1;
    }

    private int josephus(int n, int k) {
        // Base case: when there's only one person left, they are the winner
        if (n == 1) {
            return 0;
        }
        // Recursive case: calculate the position of the winner for the current state
        // josephus(n - 1, k) finds the winner position in the reduced group of (n-1)
        // people
        // (josephus(n - 1, k) + k) % n adjusts the position for the current group size
        // n
        return (josephus(n - 1, k) + k) % n;
    }
}

File: Algorithms/src/1823. Find the Winner of the Circular Game/Code/solution.cpp
class Solution
{
public:
    int findTheWinner(int n, int k)
    {
        // The problem is equivalent to the Josephus problem.
        // We need to convert the 0-indexed result from the Josephus function to 1-indexed.
        return josephus(n, k) + 1;
    }

    int josephus(int n, int k)
    {
        // Base case: when there is only one person, they are the winner.
        if (n == 1)
        {
            return 0;
        }
        // Recursive case:
        // josephus(n - 1, k) finds the position of the winner in the reduced problem (n-1 persons)
        // Adding k accounts for the step size in the problem.
        // Taking modulo n wraps around the position if it exceeds the number of people.
        return (josephus(n - 1, k) + k) % n;
    }
};
